### DISCUSSION TITLE
2022-13: 开源运营当论迹不论心

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#345] [Announcements]
# 2022-13: 开源运营当论迹不论心

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/reports/2022-13/

**Comment by sgzerolc** [comment]
感觉能够有较好纠错机制的系统才能发展地更好，社区发现问题的处理方式也能反应一些态度。

**Comment by Endle** [comment]
我觉得这篇文章写的非常好，一定程度上，帮我梳理了一下我在[发原推](https://twitter.com/ZhenboLi1/status/1511157496328687622?s=20&t=ftpPtPqcAH4TCQlJvyliEA) 时没有理清的思路。

当时我没有想清楚这个问题，就用了主观性较强的“动机”一词。现在看，动机这个词是不准确的。这篇文章中，“这个活动创造出了什么价值” 这个概念更清晰一些。

不过，我觉得这个话题可以更进一步的讨论一下。开源项目有意留下一些相对简单的 issue，让感兴趣的新人尝试，通常认为是 Good Practice. 我之前参与的 Wine 项目，很令人头疼的一点就是，很难找到难度合适的 first issue。那么，被人们称道的开源项目里的 good first issue，和灭虫活动与之的区别是什么呢？

这里，我想引入一个台阶的比喻。对于开源项目的新人[1]而言，要直接参与项目开发是非常困难的。Good first issue 就是一个小的台阶。比如很多项目的 doc fix，新人在提升项目的文档的同时，自己也认真读了至少一遍文档。如果台阶设置的过于简单，那新人就难以从中学到什么。

[1] 这里仅仅指对某个项目不了解，而不代表这个人的技术水平。就算 Graydon Hoare 对我的某个开源项目感兴趣，他也得从零开始。

[我把之前我回复 tison 的内容](https://twitter.com/ZhenboLi1/status/1510262064761159680?s=20&t=WNZRG6dCpOLxNBsLmw0zHQ) 展开一下：鼓励新人修复 lint warning，提交单元测试是好事。但是，这个工作不能做的过于简单了。 最起码，要让新人学着读一遍 dev doc，在自己本地配置好开发环境。如果这个过程中遇到了障碍，也是一个和项目其他成员的机会。这样，新人修了一个 warning，提交了几个 test case 以后，也就熟悉了基本的开发流程，可以逐步深入 code base 的挑战（深坑）了。




我写这篇评论的时候，特意看了一下 [TDengine 
good first issue](https://github.com/taosdata/TDengine/issues?q=+label%3A%22good+first+issue%22+) 到底是什么，接着发现了一个典型的例子 - <https://github.com/taosdata/TDengine/pull/8990/files> 。这个 PR 当然是值得鼓励的，但修复流程实在不敢恭维。我认为比较好的流程是，创建一个文档，标题为 “使用安全的 tstrncpy 替换不安全的 strcpy”，感兴趣的新人可以按照这个文档，学着去找出代码中有风险的 API Usage, 并提交 PR。如果创建的 issue 简单到，指出了具体哪个文件，哪一行有问题，那新人爬台阶的收益就小太多了。

从开源项目的角度看，要吸引更多的参与者，最关键的是要让参与者获得应有的收获 - 比如足够好玩，能学到东西，有解决问题的成就感等等。我衷心的希望，有更多的开源项目，无论中外，能为参与者们提供更良好的体验，而不仅仅是 “xxx项目贡献者” 这一头衔。


注：在读过 Xuanwo 的这篇博文以后，我[发了一条推](https://twitter.com/ZhenboLi1/status/1511355511089868801?s=20&t=IE5EN20naArJ-Py0Hf4ImQ) 谈了一些感想。发完以后，感觉自己的想法非常多，推特的140字远远不够，就写了一篇评论。说不定过两天我再把我的思路整理一下，就成一篇博文了 233

**Reply by Xuanwo** [reply]
> 说不定过两天我再把我的思路整理一下，就成一篇博文了 233

支持啊，欢迎出现更多的声音！

**Comment by jtao1735** [comment]
作为TDengine的创始人，认真读了这篇博客，很认同一个观点，那就是开源运营需要帮助开发者去解决真正的问题，包括TDengine的核心技术问题，而不是一些简单的typo, 对于有一定经验的程序员而言，意义不大。我们当时发起这个活动时，是因为在几个高校巡讲，想吸引在校大学生加入进来，不少学生连GitHub账号都没有，因此把门槛大大降低了一下。 

涛思数据欢迎各种批评的声音，有助于我们把各项工作做的更好。

**Reply by Xuanwo** [reply]
感谢公开回应！

想问一下 TDengine 社区除了暂停这个活动之外有更多的行动计划吗？比如引入静态检查工具在 PR 时对代码做检查等等

**Reply by tisonkun** [reply]
> 开源运营需要帮助开发者去解决真正的问题

@jtao1735 其实运营本身就有一些 marketing 的属性。“帮助”这个词太笼统了，不如说向社群成员和潜在的社群成员展示和分享 TDengine 是什么，TDengine 要解决什么问题，TDengine 是怎么解决这些问题的。从这三个角度出发，哪些地方是合适的切入点，谁是目标受众，就会清楚很多了。

每个层面的 contribution 都值得感谢，原文和我个人都不认为 fix typo 一类的工作不应该做，而是不用这样大张旗鼓的做。

> 想吸引在校大学生加入进来，不少学生连GitHub账号都没有，因此把门槛大大降低了一下。

这个说法也是不成立的。起码说明你的目的就是吸引在校大学生“进来”，也就是在 contributor graph 上出现，至于进来做什么，无所谓。这就是问题本身，而不能用来回应问题。GitHub 怎么用，首先 GitHub 自己有文档，而 PingCAP 为了解决这个问题，采取的做法是做一个系列视频[【GitHub 新手指南】极简教程教你如何给开源仓库提 PR](https://www.bilibili.com/video/BV1h5411E7pM)。

**Comment by ericyangpan** [comment]
我是 TDengine 的 DevRel 负责人，昨天读到这篇博客后，一直在整理大家反馈的问题，根据博主和 Twitter 上大家的吐槽，后面有一些事情要改进：

- [x] 关闭和清理已经 Staled 的 Issues
- ~~[ ] 基于 GitHub Actions 的 issue 清理机制 https://docs.github.com/en/github-ae@latest/actions/managing-issues-and-pull-requests/closing-inactive-issues~~
- [ ] 在研发 Workflow 中引入 https://clang.llvm.org/extra/clang-tidy/
- [ ] 尝试引入更多的工具提升代码质量，比如 https://lgtm.com/projects/g/taosdata/TDengine
- [x] 改正一些 Markdown 中历史上不正确的语法，比如 Quoting code
- [x] Good first issue 体系的重构
- [x] 后续在一些和开发者互动的过程中，用开源社区的方式去引导更多的开发者贡献代码，成为 Contributor
- [ ] 内部 Jenkins 的问题，因为历史原因，短期内可能无法改变，但长期规划上，我们会考虑以更加“开源原生”的方式处理 CI/CD 问题

TDengine 不是一个 Day one 就开源的产品，我们整个团队也是在持续摸索中，无论是研发、CI、社区还是运营。但既然选择了开源，一定会坚定的拥抱开源社区，至少做到 @Phodal 提及的 Useful Open Source 级开源水平：https://dx.phodal.com/docs/anti-patterns/fake-opensource.html 我们也保持着开放的心态，向开源社区学习，虚心听取来自社区开发者的意见。

**Reply by Xuanwo** [reply]
很赞的行动计划！另外建议在 TDengine 创建一个独立的 tracking issue 以追踪这些改进，这样社区也可以加入进去 :rocket:

**Reply by ericyangpan** [reply]
目前 TDengine 的 Workflow 还没做到 GitHub Issue Lead，需要稍微多一点更多的时间去推动这个事情的实践，尽快

> 很赞的行动计划！另外建议在 TDengine 创建一个独立的 tracking issue 以追踪这些改进，这样社区也可以加入进去 🚀

**Reply by tisonkun** [reply]
@arielyang 我看前两个都是跟 stale 相关的问题，stale 这件事我是不建议走机器人的。反正从我的经历和经验上看，走机器人会出一大堆 false positive 而且很烦人。

> 我是严重反对 stale bot 这种东西的。
>
> 要么是走 zeromq / skywalking 的思路，几天没人理就 close 掉。要么就放着。依靠机器人只会降低消息流通的质量。
>
> 因为没人为机器人的行为负责。

**Reply by tisonkun** [reply]
> 在研发 Workflow 中引入 https://clang.llvm.org/extra/clang-tidy/

@arielyang 我搜了一下 taosdata/TDengine 的仓库，没有找到对应的 issue 耶。这个直接建个 issue 就行了。还是说典中典已经有个内部 JIRA ticket 就好了 =。=

**Reply by tisonkun** [reply]
>  Good first issue 体系的重构

现在也没几个 good first issue，就是上次整蛊以后留下来四个月没啥进展的 issue，有人力的话重新评估一下顺手处理掉或者关掉就算了。good first issue 正常来就行，一般来说就是你认为公司实习生能 pick 就打这个标签，至于谁解其实从 community 的角度看没啥区别，不是说 good first issue senior 就不能解了。

**Reply by tisonkun** [reply]
> * 尝试引入更多的工具提升代码质量，比如 https://lgtm.com/projects/g/taosdata/TDengine
> * 改正一些 Markdown 中历史上不正确的语法，比如 Quoting code
> * 后续在一些和开发者互动的过程中，用开源社区的方式去引导更多的开发者贡献代码，成为 Contributor
> * 内部 Jenkins 的问题，因为历史原因，短期内可能无法改变，但长期规划上，我们会考虑以更加“开源原生”的方式处理 CI/CD 问题

都是很正确的话，期待后续发文跟进进展。不过第一点我持保留意见，加工具不代表代码质量就变好了。

**Reply by ericyangpan** [reply]
@tisonkun 感谢这么多反馈，感受到了社区的力量。
第一，关于 stale 的问题，我认为“几天没人理就 close 掉”也是一个规则，实际上还是有活跃的社区成员执行。我是希望有 bot 去做兜底的规则，比如第 30 天 comment，第 45 天 close。就像我们做 GTD 一样，如果一个 issue 被关闭是不可控的，这个动作要么可能不被执行，要么可能花费更多的成本执行，这两点我都不希望发生。
第二，关于 create issue 的事情，上周内部已经有一些推动，但是目前开发体系还没完全切换到 GitHub Issues 上，担心 workflow 会乱掉，也需要和相关同事达成充分的沟通和共识，所以我这边还需要一些时间。
